Add basic tests for mixin subverbs

leander-dsouza opened the pull request:
## Basic Info

| Info | Result |
| ------ | ----------- |
| Primary OS tested on | Ubuntu |

## Description of contribution in a few bullet points

* This PR adds base integration tests for colcon mixin subverbs, covering the following:

   * Adding mixin repositories
   * Updating mixin repositories
   * Showing available mixins
   * Removing mixin repositories
   * Error handling for invalid inputs and duplicate operations
  
* Included a couple of mixin files - `build-type.mixin` and `coverage.mixin` from the [colcon-mixin-repository](https://github.com/colcon/colcon-mixin-repository) to serve as a basis for all repository testing.

## Description of how this change was tested

* Ensured all the tests pass locally - `python3 -m pytest test`.

leander-dsouza committed:
Added base colcon mixin files.

Signed-off-by: Leander Stephen D'Souza <leanderdsouza1234@gmail.com>

leander-dsouza committed:
Added basic tests for colcon mixin subverbs.

Signed-off-by: Leander Stephen D'Souza <leanderdsouza1234@gmail.com>

cottsay commented:
Thanks for making some progress on the tests.

Invoking the entirety of colcon as part of the tests here isn't something we'll want to do. There are a variety of reasons, but to name a couple:
1. The developer may have any number of extensions enabled, some of which may interfere with the test expectations.
2. The in-test invocations could cause other extensions to act unpredictably, such as `colcon-rerun`, which records invocations.

Other extensions test their verbs without using subprocess invocations, and we should do the same here. Examples:
* https://github.com/colcon/colcon-alias/blob/main/test/test_verb_alias.py
* https://github.com/colcon/colcon-core/blob/master/test/test_verb_build.py

leander-dsouza commented:
> Invoking the entirety of colcon as part of the tests here isn't something we'll want to do.

Thank you for attaching the documentation of the testing pattern of other members of `colcon`.
I have updated the tests to use `CommandContext` instead of `subprocess` to validate each subverb usage.

cottsay commented on test/test_subverb.py:
```diff
@@ -0,0 +1,167 @@
+set_default_config_path(path=Path('/tmp/colcon_mixin_test'))
+
+# Import colcon-mixin modules after config path is set
+# flake8: noqa: E402
```
This is a pretty poorly implemented facet of `colcon-mixin`. Let's fix it so that this isn't necessary. Mind taking a look at #55?

leander-dsouza replied:
I have reviewed the PR; however, we still need to call `set_default_config_path`, as if this is not set, importing `get_config_path` would lead to an [`AssertionError`](https://github.com/colcon/colcon-core/blob/master/colcon_core/location.py#L30).

leander-dsouza replied:
I am closing this conversation, as the importing issue with `set_default_config_path` has been resolved.

cottsay commented on test/expected_mixins/demo.txt:
We should align the YAML style with the rest of the project. Mind taking a look at #56, which adds a check for this?

leander-dsouza replied:
The [show module](https://github.com/colcon/colcon-mixin/blob/master/colcon_mixin/subverb/show.py#L74-L78) of colcon-mixin does not print the mixins in a standardised YAML or JSON format. Hence, I cannot change this to an expected YAML output.

cottsay replied:
Whoa, bummer. That might be worth filing a new issue.

leander-dsouza replied:
I have created #58 to ensure that the `show` subverb would print YAML output.
If you don't mind, could you please take a look at it?

Upon merging the said PR, I will modify `demo.txt` to have an expected YAML output.

cottsay commented on test/mixins/build-type.mixin:
Unless there's a reason to use JSON here, I think we should use YAML. For one, it has better line-change-cleanliness. The filename can remain the same or change, whatever you prefer.

leander-dsouza replied:
I have modified the mixins to use a YAML format instead.
However, `yamllint` cannot check for linting in these files, as these have to end in a `.mixin` extension for [discovery](https://github.com/colcon/colcon-mixin/blob/master/colcon_mixin/mixin/__init__.py#L66).

cottsay reviewed (changes requested):
Thanks, this is looking a lot better. Looking forward to getting it merged and seeing that coverage number go up.

leander-dsouza committed:
Adhere mixins to a YAML format.

Signed-off-by: Leander Stephen D'Souza <leanderdsouza1234@gmail.com>

cottsay commented on test/test_subverb.py:
```diff
@@ -0,0 +1,167 @@
+
+    assert result is None
+
+    expected_demo_file = Path(__file__).parent.joinpath('expected_mixins', 'demo.txt')
```
A more common way to combine path atoms, which also happens to be compact enough to fit this line under the 80 column limit.

```suggestion
    expected_demo_file = Path(__file__).parent / 'expected_mixins' / 'demo.txt'
```

cottsay replied:
Merging this to fix CI

cottsay committed:
Merge branch 'master' into test-colcon-mixin-add

cottsay committed:
Drop import hacks for get_config_path

This is no longer needed after 9f2dc674a51911e74fb735ddb1017a6913d548bc
was merged.

Additionally, use a pytest temporary directory for staging the config
files rather than a static location.

cottsay committed:
Shorten a long line

cottsay commented:
I merged master and added [a commit](https://github.com/colcon/colcon-mixin/pull/54/commits/d09b50f1fb5bacb6d3ad99cd3f177d6ba378b225) to remove the special casing around `get_config_path` now that #55 is merged.
